From a08c272056b4f6228bbae1eb56d94447867de594 Mon Sep 17 00:00:00 2001
From: Purushottam Narayana <purushottam.narayana@lge.com>
Date: Wed, 10 Jan 2018 14:32:49 -0800
Subject: [PATCH] Fix scaling not working for specific display windows

:Release Notes:
Fix scaling not working for specific display window.

:Detailed Notes:
Some of the coordinates considered invalid seems to be valid ones in YT.

:Testing Performed:
Play youtube in webbrowser and confirm it fits correctly.

:QA Notes:

:Issues Addressed:
[PLAT-48284] Video is not displayed well on youtube site in web browser app
---
 drivers/gpu/drm/drm_crtc.c | 28 ++++++++++++++--------------
 1 file changed, 14 insertions(+), 14 deletions(-)

diff --git a/drivers/gpu/drm/drm_crtc.c b/drivers/gpu/drm/drm_crtc.c
index c149aee..5028ae7 100644
--- a/drivers/gpu/drm/drm_crtc.c
+++ b/drivers/gpu/drm/drm_crtc.c
@@ -2401,7 +2401,7 @@ static int __setplane_internal(struct drm_plane *plane,
 	}
 
 	/* BRCM RPI drivers gives div 0 exception when display width/height being 0 */
-	if (!(crtc_w & crtc_h)) {
+	if (crtc_w == 0 || crtc_h == 0) {
 		DRM_DEBUG_KMS("Invalid CRTC coordinates %ux%u+%d+%d\n",
 			      crtc_w, crtc_h, crtc_x, crtc_y);
 		ret = -ERANGE;
@@ -2444,15 +2444,13 @@ static int setplane_internal(struct drm_plane *plane,
 			     uint32_t src_w, uint32_t src_h)
 {
 	int ret;
-        DRM_DEBUG_KMS("\n setplane_internal ENTER");
-
-
+	DRM_DEBUG_KMS("\n setplane_internal ENTER");
 	drm_modeset_lock_all(plane->dev);
 	ret = __setplane_internal(plane, crtc, fb,
 				  crtc_x, crtc_y, crtc_w, crtc_h,
 				  src_x, src_y, src_w, src_h);
 	drm_modeset_unlock_all(plane->dev);
-        DRM_DEBUG_KMS("\n setplane_internal EXIT");
+	DRM_DEBUG_KMS("\n setplane_internal EXIT");
 
 	return ret;
 }
@@ -5002,7 +5000,7 @@ int drm_mode_plane_set_custom_prop(struct drm_device *dev, struct drm_plane *pla
 		uint32_t src_h, src_w;
 	} *scale_param;
 
-	DRM_DEBUG_KMS("\n drm_mode_plane_set_custom_prop arg->value = %llx arg->prop_id = %x plane_id = %d",arg->value, arg->prop_id, arg->obj_id);
+	DRM_DEBUG_KMS("drm_mode_plane_set_custom_prop arg->value = %llx arg->prop_id = %x plane_id = %d \n",arg->value, arg->prop_id, arg->obj_id);
 
 	switch (arg->prop_id) {
 		case 0xff02:
@@ -5051,18 +5049,20 @@ int drm_mode_plane_set_custom_prop(struct drm_device *dev, struct drm_plane *pla
 			return ret;
 		case 0xff03:
 			scale_param = (struct io_rect *) arg->value;
-			DRM_DEBUG_KMS("src rect: %d %d %d %d", scale_param->src_x, scale_param->src_y,
+			DRM_DEBUG_KMS("src rect: %d %d %d %d \n", scale_param->src_x, scale_param->src_y,
 				   scale_param->src_w, scale_param->src_h);
-			DRM_DEBUG_KMS("crtc rect: %d %d %d %d", scale_param->crtc_x, scale_param->crtc_y,
+			DRM_DEBUG_KMS("crtc rect: %d %d %d %d \n", scale_param->crtc_x, scale_param->crtc_y,
 				   scale_param->crtc_w, scale_param->crtc_h);
 
 			/* BRCM RPI drivers gives div 0 exception when display width/height being 0 */
-			if (!scale_param || !(scale_param->crtc_w & scale_param->crtc_h)) {
+			/* Scale param check is not required anymore as stack address passing is rectified in
+			 upper layer */
+			if (scale_param->crtc_w == 0 || scale_param->crtc_h == 0) {
 				DRM_DEBUG_KMS("\n Invalid values, Dst rect width or height or both zeros");
 				return -ENOENT;
 			}
 
-			if (!fb) {
+			if (!plane->state->fb) {
 				DRM_DEBUG_KMS("\n No FB connected yet, store the values");
 				plane->state->crtc_x = scale_param->crtc_x;
 				plane->state->crtc_y = scale_param->crtc_y;
@@ -5073,13 +5073,13 @@ int drm_mode_plane_set_custom_prop(struct drm_device *dev, struct drm_plane *pla
 				plane->state->src_y = scale_param->src_y;
 				plane->state->src_w = scale_param->src_w << 16;
 				plane->state->src_h = scale_param->src_h << 16;
+				ret = 0;
 			} else {
-				ret = __setplane_internal_custom(plane, crtc, fb, scale_param->crtc_x, scale_param->crtc_y,
+				ret = __setplane_internal_custom(plane, crtc, plane->state->fb, scale_param->crtc_x, scale_param->crtc_y,
 												 scale_param->crtc_w, scale_param->crtc_h,
 												 scale_param->src_x, scale_param->src_y,
 												 scale_param->src_w << 16, scale_param->src_h << 16);
 			}
-
 			break;
 
 		case 0xff01:
@@ -5103,12 +5103,12 @@ int drm_mode_plane_set_custom_prop(struct drm_device *dev, struct drm_plane *pla
 				else
 					return -ENOENT;
 			}
-			if (!(plane->state->crtc_w & plane->state->crtc_h)) {
+			if (plane->state->crtc_w == 0 || plane->state->crtc_h == 0) {
 				plane->state->crtc_w = fb->width;
 				plane->state->crtc_h = fb->height;
 				DRM_DEBUG_KMS("\n Dst width or height of dst is 0");
 			}
-			if (!(plane->state->src_w & plane->state->src_h)) {
+			if (plane->state->src_w == 0 || plane->state->src_h == 0) {
 				/* src_{x,y,w,h} values are 16.16 fixed point */
 				plane->state->src_w = fb->width << 16;
 				plane->state->src_h = fb->height << 16;
-- 
1.9.1

